File: WFInfo/LogCapture.vb
Imports System
Imports System.IO
Imports System.IO.MemoryMappedFiles
Imports System.Security.AccessControl
Imports System.Security.Principal
Imports System.Text
Imports System.Threading

Public Class LogCapture
    Implements IDisposable
    Private memoryMappedFile As MemoryMappedFile = Nothing
    Private bufferReadyEvent As EventWaitHandle = Nothing
    Private dataReadyEvent As EventWaitHandle = Nothing
    Private tokenSource As New CancellationTokenSource()
    Private token As CancellationToken = tokenSource.Token

    Public Sub New()
        Main.addLog("STARTING LogCapture")
        memoryMappedFile = MemoryMappedFile.CreateOrOpen("DBWIN_BUFFER", 4096L)
        Dim created As Boolean

        bufferReadyEvent = New EventWaitHandle(False, EventResetMode.AutoReset, "DBWIN_BUFFER_READY", created)

        If Not created Then
            Main.addLog("The DBWIN_BUFFER_READY event exists.")
            Return
        End If

        dataReadyEvent = New EventWaitHandle(False, EventResetMode.AutoReset, "DBWIN_DATA_READY", created)

        If Not created Then
            Main.addLog("The DBWIN_DATA_READY event exists.")
            Return
        End If

        Task.Factory.StartNew(AddressOf Run)
    End Sub

    Public Sub Run()
        Dim proc As Process = parser2.GetWFProc()
        Try
            Dim Timeout = TimeSpan.FromSeconds(1.0)
            bufferReadyEvent.Set()
            While Not token.IsCancellationRequested

                If Not dataReadyEvent.WaitOne(Timeout) Then
                    Continue While
                End If
                If proc Is Nothing OrElse proc.HasExited Then
                    proc = parser2.GetWFProc()
                End If
                If proc IsNot Nothing Then
                    Using stream = memoryMappedFile.CreateViewStream()
                        Using reader = New BinaryReader(stream, Encoding.Default)
                            Dim processId = reader.ReadUInt32()
                            If processId = proc.Id Then
                                Dim chars = reader.ReadChars(4092)
                                Dim index = Array.IndexOf(chars, Chr(0))
                                Dim message = New String(chars, 0, index)
                                RaiseEvent TextChanged(Me, message.Trim)
                            End If
                        End Using
                    End Using
                End If

                ' The message has been processed, so trigger the
                ' DBWIN_BUFFER_READY event in order to receive the next message
                ' from the process being debugged.
                bufferReadyEvent.Set()
            End While
        Catch ex As Exception
            Console.WriteLine(ex.ToString())
        Finally
            If memoryMappedFile IsNot Nothing Then
                memoryMappedFile.Dispose()
            End If

            If bufferReadyEvent IsNot Nothing Then
                bufferReadyEvent.Dispose()
            End If

            If dataReadyEvent IsNot Nothing Then
                dataReadyEvent.Dispose()
            End If
        End Try
    End Sub

    Public Sub Dispose() Implements IDisposable.Dispose
        tokenSource.Cancel()
        tokenSource.Dispose()
        Main.addLog("STOPPING LogCapture")
    End Sub

    Public Delegate Sub LogWatcherEventHandler(sender As Object, text As String)

    ''Event that is fired when the file is changed
    Public Event TextChanged As LogWatcherEventHandler
End Class

File: WFInfo/LogCapture.cs
using System;
using System.Diagnostics;
using System.IO;
using System.IO.MemoryMappedFiles;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace WFInfo
{
    public delegate void LogWatcherEventHandler(object sender, String text);

    class LogCapture : IDisposable
    {
        private readonly MemoryMappedFile memoryMappedFile;
        private readonly EventWaitHandle bufferReadyEvent;
        private EventWaitHandle dataReadyEvent;
        readonly CancellationTokenSource tokenSource = new CancellationTokenSource();
        private CancellationToken token;
        private readonly Timer timer;
        public event LogWatcherEventHandler TextChanged;

        public LogCapture()
        {
            token = tokenSource.Token;
            Main.AddLog("Starting LogCapture");
            memoryMappedFile = MemoryMappedFile.CreateOrOpen("DBWIN_BUFFER", 4096L);

            bufferReadyEvent = new EventWaitHandle(false, EventResetMode.AutoReset, "DBWIN_BUFFER_READY", out Boolean createdBuffer);

            if (!createdBuffer)
            {
                Main.AddLog("The DBWIN_BUFFER_READY event exists.");
                return;
            }

            var startTimeSpan = TimeSpan.Zero;
            var periodTimeSpan = TimeSpan.FromSeconds(10);

            timer = new Timer((e) =>
            {
                GetProcess();
            }, null, startTimeSpan, periodTimeSpan);

        }

        private void Run()
        {

            try
            {
                TimeSpan timeout = TimeSpan.FromSeconds(1.0);
                bufferReadyEvent.Set();
                while (!token.IsCancellationRequested)
                {


                    if (!dataReadyEvent.WaitOne(timeout))
                    {
                        continue;
                    }

                    if (OCR.OCR.Warframe != null)
                    {
                        using (MemoryMappedViewStream stream = memoryMappedFile.CreateViewStream())
                        {
                            using (BinaryReader reader = new BinaryReader(stream, Encoding.Default))
                            {
                                uint processId = reader.ReadUInt32();
                                if (processId == OCR.OCR.Warframe.Id)
                                {
                                    char[] chars = reader.ReadChars(4092);
                                    int index = Array.IndexOf(chars, '\0');
                                    string message = new string(chars, 0, index);
                                    TextChanged(this, message.Trim());
                                }
                            }
                        }
                    }
                    bufferReadyEvent.Set();
                }
            }
            catch (Exception ex)
            {
                Main.AddLog(ex.ToString());
                Main.RunOnUIThread(() =>
                {
                    _ = new ErrorDialogue(DateTime.Now, 0);
                });
            }
            finally
            {
                if (memoryMappedFile != null)
                    memoryMappedFile.Dispose();

                if (bufferReadyEvent != null)
                    bufferReadyEvent.Dispose();

                if (dataReadyEvent != null)
                    dataReadyEvent.Dispose();
            }
        }

        private void GetProcess()
        {
            if ((OCR.OCR.Warframe == null) || (OCR.OCR.Warframe.HasExited))
            {
                if (!OCR.OCR.VerifyWarframe())
                    return;
            }
            dataReadyEvent = new EventWaitHandle(false, EventResetMode.AutoReset, "DBWIN_DATA_READY", out Boolean createdData);

            if (!createdData)
            {
                Main.AddLog("The DBWIN_DATA_READY event exists.");
                return;
            }

            Task.Factory.StartNew(Run);
            timer.Dispose();
        }

        public void Dispose()
        {
            if (memoryMappedFile != null)
                memoryMappedFile.Dispose();

            if (bufferReadyEvent != null)
                bufferReadyEvent.Dispose();

            if (dataReadyEvent != null)
                dataReadyEvent.Dispose();

            tokenSource.Cancel();
            tokenSource.Dispose();
            Main.AddLog("Stoping LogCapture");
        }
    }
}
